writeaheadlog.go

```diff
@@ -226,29 +283,47 @@ func (w *WAL) recoverWal(data []byte) ([]Update, error) {
 	return updates, nil
 }
 
+// writeRecoveryState is a helper function that changes the recoveryState on disk
+func (w *WAL) writeRecoveryState(state uint16) error {
+	// Set the metadata to unclean again
+	var recoveryState []byte
+	binary.LittleEndian.PutUint16(recoveryState, recoveryStateUnclean)
+	if _, err := w.logFile.WriteAt(recoveryState, 0); err != nil {
+		return err
+	}
+	return nil
+}
+
 // RecoveryComplete is called after a wal is recovered to signal that it is
 // safe to reset the wal
 func (w *WAL) RecoveryComplete() error {
-	// Marshal the pageStatusApplied
-	pageAppliedBytes := make([]byte, 8)
-	binary.LittleEndian.PutUint64(pageAppliedBytes, pageStatusApplied)
+	// Set the metadata to swipe
+	if err := w.writeRecoveryState(recoveryStateSwipe); err != nil {
+		return err
+	}
 
-	// Get the length of the file.
-	length, err := w.logFile.Seek(0, io.SeekEnd)
-	if err != nil {
+	// Sync before we start wiping
+	if err := w.logFile.Sync(); err != nil {
 		return err
 	}
 
-	// Set all pages to applied.
-	for offset := int64(pageSize + checksumSize); offset < length; offset += pageSize {
-		if _, err := w.logFile.WriteAt(pageAppliedBytes, offset); err != nil {
-			return err
-		}
+	// Wipe the wal
+	if err := w.wipeWAL(); err != nil {
+		return err
+	}
+
+	// Set the metadata to unclean again
+	if err := w.writeRecoveryState(recoveryStateUnclean); err != nil {
```

Comment:
Need to sync before you set the recovery state to unclean. Then you need to sync again.

Reply:
So basically, recovery is going to have 3 serial syncs:

1. after changing metadata to wiping
2. after wiping
3. after changing metadata to unclean